crates/ide_assists/src/handlers/generate_documentation_template.rs

```diff
@@ -0,0 +1,1025 @@
+            lines.push(format!("assert_eq!({}, );", ex_helper.function_call));
+        } else {
+            lines.push(format!("let result = {};", ex_helper.function_call));
+            lines.push("assert_eq!(result, );".into());
+        }
+    } else {
+        lines.push(format!("{};", ex_helper.function_call));
+    }
+    // Check the mutated values
+    if is_ref_mut_self(ast_func) == Some(true) {
+        lines.push(format!("assert_eq!({}, );", ex_helper.self_name?));
+    }
+    for param_name in &ex_helper.ref_mut_params {
+        lines.push(format!("assert_eq!({}, );", param_name));
+    }
+    Some(lines)
+}
+
+/// Generate an example template which should panic
+/// `None` if the function has a `self` parameter but is not in an `impl`.
+fn gen_panic_ex_template(ast_func: &ast::Fn, krate_name: String) -> Option<Vec<String>> {
+    let (mut lines, ex_helper) = gen_ex_start_helper(ast_func, krate_name)?;
+    match returns_a_value(ast_func) {
+        true => lines.push(format!("let _ = {}; // panics", ex_helper.function_call)),
+        false => lines.push(format!("{}; // panics", ex_helper.function_call)),
+    }
+    Some(lines)
+}
+
+/// Intermediary results of the start of example generation
+struct ExHelper {
+    function_call: String,
+    param_list: ast::ParamList,
+    ref_mut_params: Vec<String>,
+    self_name: Option<String>,
+}
+
+/// Build the start of the example and transmit the useful intermediary results.
+/// `None` if the function has a `self` parameter but is not in an `impl`.
+fn gen_ex_start_helper(ast_func: &ast::Fn, krate_name: String) -> Option<(Vec<String>, ExHelper)> {
+    let mut lines = Vec::new();
+    let is_unsafe = ast_func.unsafe_token().is_some();
+    let param_list = ast_func.param_list()?;
+    let ref_mut_params = ref_mut_params(&param_list);
+    let self_name: Option<String> = self_name(ast_func);
+
+    lines.push(format!("use {};", build_path(ast_func, krate_name)));
+    lines.push("".into());
+    if let Some(self_definition) = self_definition(ast_func, self_name.as_deref()) {
+        lines.push(self_definition);
+    }
+    for param_name in &ref_mut_params {
+        lines.push(format!("let mut {} = ;", param_name))
+    }
+    let function_call = function_call(ast_func, &param_list, self_name.as_deref(), is_unsafe)?;
+    let ex_helper = ExHelper { function_call, param_list, ref_mut_params, self_name };
+    Some((lines, ex_helper))
+}
+
+/// Check if the function and all its parent modules are exactly `pub`
+fn is_public(ast_func: &ast::Fn) -> bool {
+    has_pub(ast_func)
+        && ast_func
+            .syntax()
+            .ancestors()
+            .filter_map(ast::Module::cast)
+            .all(|module| has_pub(&module))
+}
+
+/// Check if visibility is exactly `pub` (not `pub(crate)` for instance)
+fn has_pub<T: HasVisibility>(item: &T) -> bool {
+    item.visibility().map(|v| v.path().is_none()).unwrap_or(false)
+}
```

Comment:
I think a better approach here is to turn the `ast::Fn` into a `hir::Fn` via `sema.to_def` and then using `hir::HasVisibility` with the crate's root module to check if its visible from the root or whever. The current impl only checks the file this is in as its only using the syntax nodes.

Reply:
> (I am discovering pub in trait is not interesting (removed when formatting) but needed to apply the assist

Note that this is an error by default for rustc even. With this change you wouldn't need to special case trait functions not having pub either I believe

Reply:
Is it what you meant?

```rust
fn is_public(ast_func: &ast::Fn, ctx: &AssistContext) -> Option<bool> {
    let visibility = ast_to_hir_fn(ast_func, ctx)?.visibility(ctx.db());
    Some(visibility == Visibility::Public)
}

fn ast_to_hir_fn(ast_func: &ast::Fn, ctx: &AssistContext) -> Option<ModuleDef> {
    Some(ctx.sema.to_def(ast_func)?.module(ctx.db()).into())
}
```

It seems to work during manual tests but it does not work during automated tests… In the automated tests `dbg!()` indicates that

```rust
visibility = Module(
    ModuleId {
        krate: CrateId(
            0,
        ),
        block: None,
        local_id: Idx::<ModuleData>(0),
    },
)
```

Reply:
Okay, this time the code seems good but it does not work:

```rust
fn is_public(ast_func: &ast::Fn, ctx: &AssistContext) -> Option<bool> {
    let hir_func = ctx.sema.to_def(ast_func)?;
    Some(hir_func.visibility(ctx.db()) == Visibility::Public)
}
```

It results in

```
test handlers::generate_documentation_template::tests::checks_ref_mut_in_example ... ok
test handlers::generate_documentation_template::tests::checks_modified_self_param ... ok
test handlers::generate_documentation_template::tests::checks_output_in_example ... ok
test handlers::generate_documentation_template::tests::detects_new ... ok
test handlers::generate_documentation_template::tests::detects_result_output ... ok
test handlers::generate_documentation_template::tests::guesses_assert_macro_can_panic ... ok
test handlers::generate_documentation_template::tests::detects_unsafe_function ... ok
test handlers::generate_documentation_template::tests::guesses_panic_macro_can_panic ... ok
test handlers::generate_documentation_template::tests::guesses_expect_can_panic ... ok
test handlers::generate_documentation_template::tests::detects_new_with_self ... ok
test handlers::generate_documentation_template::tests::guesses_unwrap_can_panic ... ok
test handlers::generate_documentation_template::tests::not_applicable_if_function_already_documented ... ok
test handlers::generate_documentation_template::tests::not_applicable_if_function_is_private ... ok
test handlers::generate_documentation_template::tests::not_applicable_if_function_is_pub_crate ... ok
test handlers::generate_documentation_template::tests::not_applicable_in_trait_impl ... ok
test handlers::generate_documentation_template::tests::not_applicable_if_function_is_in_private_mod ... FAILED
test handlers::generate_documentation_template::tests::not_applicable_on_function_calls ... ok
test handlers::generate_documentation_template::tests::not_applicable_if_function_is_in_non_public_mod_is_recursive ... FAILED
test handlers::generate_documentation_template::tests::not_applicable_if_function_is_in_pub_crate_mod ... FAILED
test handlers::generate_documentation_template::tests::supports_fn_in_mods ... ok
test handlers::generate_documentation_template::tests::supports_a_parameter ... ok
test handlers::generate_documentation_template::tests::stores_result_if_at_least_3_params ... ok
test handlers::generate_documentation_template::tests::supports_fn_in_impl ... ok
test handlers::generate_documentation_template::tests::supports_method_call ... ok
test handlers::generate_documentation_template::tests::supports_noop_function ... ok
test tests::generated::doctest_generate_documentation_template ... ok
```

So there are three failed tests

```
test handlers::generate_documentation_template::tests::not_applicable_if_function_is_in_private_mod ... FAILED
test handlers::generate_documentation_template::tests::not_applicable_if_function_is_in_non_public_mod_is_recursive ... FAILED
test handlers::generate_documentation_template::tests::not_applicable_if_function_is_in_pub_crate_mod ... FAILED
```

In the three failed tests `visibility` is `Visibility::Public`, which is documented as unrestricted visibility. :thinking:

Reply:
Ye this only applies to the visibility of the item, it does not check if the item is properly exported from the crate, which I think is fine here.

Reply:
Thanks for the answer! My concern here is that there will be a generated example template; the author will take time to fill it but it will not build because the function is not exported. Maybe using this property + using the `hir` to find all the parent `mod`s (not only in the current file, contrary to the `ast` version iiuc) to check that they are all public would be better ?

Reply:
There are lints against having public items that are unreachable(I think even r-a has that enabled on CI)
But otherwise, I think you should be able to fetch the krate of the function it is defined is, fetch the root module and query whether its visible from there by asking for an importable path with `hir::Module::find_use_path` I think.

Reply:
Before you published this comment, I wrote the function below which works in both the automated tests and the manual tests.

```rust
fn all_parent_mods_public(hir_func: &hir::Function, ctx: &AssistContext) -> bool {
    let mut module = hir_func.module(ctx.db());
    loop {
        if let Some(parent) = module.parent(ctx.db()) {
            match ModuleDef::from(module).visibility(ctx.db()) {
                Visibility::Public => module = parent,
                _ => break false,
            }
        } else {
            break true;
        }
    }
}
```

Then I read the comment, I tried using `hir::Module::find_use_path` and it seems to return `Some` path even if there is something private in the path.